packages/nitro-protocol/contracts/interfaces/IMultiAssetHolder.sol

```diff
@@ -2,50 +2,12 @@
 pragma solidity 0.7.4;
 pragma experimental ABIEncoderV2;
 
-import '../Outcome.sol';
+import {ExitFormat as Outcome} from '@statechannels/exit-format/contracts/ExitFormat.sol';
 
 /**
  * @dev The IMultiAssetHolder interface calls for functions that allow assets to be transferred from one channel to other channel and/or external destinations, as well as for guarantees to be claimed.
  */
 interface IMultiAssetHolder {
-    /**
-     * @notice Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
-     * @dev Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
-     * @param assetIndex The index of the asset to be paid out (in the outcome)
-     * @param fromChannelId Unique identifier for state channel to transfer funds *from*.
-     * @param outcomeBytes The abi.encode of AssetOutcome.Allocation
-     * @param stateHash the hash of the state stored on chain
-     * @param indices Array with each entry denoting the index of a destination to transfer funds to.
-     */
-    function transfer(
-        uint256 assetIndex,
-        bytes32 fromChannelId,
-        bytes calldata outcomeBytes,
-        bytes32 stateHash,
-        uint256[] memory indices
-    ) external;
-
-    /**
-     * @notice Transfers as many funds escrowed against `guarantorChannelId` as can be afforded for a specific destination in the beneficiaries of the __target__ of that channel. Checks against the storage in this contract.
-     * @dev Transfers as many funds escrowed against `guarantorChannelId` as can be afforded for a specific destination in the beneficiaries of the __target__ of that channel. Checks against the storage in this contract.
-     * @param assetIndex The index of the asset to be paid out (in the outcome)
-     * @param guarantorChannelId Unique identifier for a guarantor state channel.
-     * @param guarantorOutcomeBytes The abi.encode of the guarantor outcome.
-     * @param guarantorStateHash the hash of the state stored on chain for the guarantor
-     * @param targetOutcomeBytes The abi.encode of the guarantor outcome.
-     * @param targetStateHash the hash of the state stored on chain for the guarantor
-     * @param indices Array with each entry denoting the index of a destination (in the target channel) to transfer funds to. Should be in increasing order. An empty array indicates "all".
-     */
-    function claim(
-        uint256 assetIndex,
-        bytes32 guarantorChannelId,
-        bytes memory guarantorOutcomeBytes,
-        bytes32 guarantorStateHash,
-        bytes memory targetOutcomeBytes,
-        bytes32 targetStateHash,
-        uint256[] memory indices
-    ) external;
```

Comment:
@NiloCK I think we want something like the above API for `claim`. 

Firstly, about the `pure` declaration. We do want a pure function at the "heart" of claim -- but fundamentally the top level claim method needs to read from and mutate the state of the chain. So what you have now should become `_compute_claim_effects_and_interactions`, and we should then wrap it with inside a `claim` that has the above API. We will also then need a `_apply_claim_checks` and a `_apply_claim_effects` as well as `_apply_claim_interactions` so that we mirror the factorization of `transfer` that we have on this branch.

Secondly, about the inclusion of `assetIndex` in the API. I think we want it! We do not want to loop over all assets  (see https://github.com/statechannels/exit-format/issues/48). Although this is what we currently do in the exit format repo, looping over all assets doesn't really make sense with only a single `targetChannelIndex` -- it means all allocations in the source (aka guaranteeing) channel need to be present in every asset slice of its outcome. What I would like to do in the short term is match the current API of nitro protocol -- that is to say, slice into a single asset with `assetIndex`, and then only provide a 1d array of `indices` with which to select allocations from the target to pay out. I think you should find something pretty much exactly what I have in mind on this PR https://github.com/statechannels/exit-format/pull/52 under the guise of `_computeNewAllocationWithGuarantee`. I think that should be pasted here and become the `_compute_claim_effects_and_interactions`.

Reply:
I have been thinking about the second point for a bit. When `claim` is invoked, the guarantor and the channels it targets are finalized. At that point, it is in the interest of the guarantor participants to move all funds from the guarantor channel to the target channel(s). So `claim` would get invoked with every asset that contains funds for the guarantor channel.

If the above point make sense, then it seems like we are including the `assetIndex` as a consequence of how asset-to-channel data is stored. I wonder if there is a data structure we could use to enable the functionality to claim all assets without the restriction that every asset must contain every guarantor channel.

Reply:
One alternative that comes to mind is specifying the `targetChannelId` in the `claim` API. With this API, we need to traverse every allocation for every asset. We would then check whether the the allocation destination matches the `targetChannelId`. This is probably too inefficient.

Reply:
I think the overall data structures and claim API are all fair game for improvement. In the short term however, the goal is to bring in the current exit format. This is so we can benefit from it allowing heterogenous outcomes (and also several guarantees in one outcome) and move away from having dedicated guarantor channels. 

I think a good strategy for now is to keep the API for `claim` as close to the current API as possible. Seeing as the current API targets a single guarantee and a single asset, we will need to add `targetChannelIndex` or `targetChannelId` to the API -- given that a ledger channel can contain many guarantees. As you say, `targetChannelId` would force us to do more "searching" through the outcomes -- and give us even more nested loops. So think sticking with an index makes sense for now. 


My attempt at `claim` is over at #3714 .